### PR TITLE
[week5] - jungmin

### PR DESCRIPTION
**Pull request by lego37yoon**
# 그래디언트 클리핑

> Gradient Clipping

그래디언트 폭주(exploding gradient) 문제를 줄이기 위한 방법으로, 역전파 단계에서 그래디언트 값이 특정 임계값(threshold)을 넘지 않도록 잘라낸다.

# 심층 신경망 재사용

규모가 매우 큰 DNN 모델 학습 시 새로 학습하는 작업은 느리다. 이때 기존 학습된 모델이 존재하면 하위 층(lower layer)을 가져와 재사용하는 것이 학습 속도를 빠르게 하고 학습에 필요한 데이터 세트도 적다. 이러한 방법을 Transfer Learning이라고 한다.

이때 재사용에 사용하지 않을 층은 동결(freezing)하고, 사용할 층은 학습 가능한 상태로 유지한다. 동결된 층은 학습 과정에서 변하지 않으므로 가장 마지막에 동결된 층 출력을 캐싱하여 활용하면 학습 속도가 개선된다.

# 손실함수

모델의 예측 값과 실제 값 사이읭 오차를 계산한다.

## MSE

> Mean Squared Error

실제 값과 예측 값 사이 차이를 오차제곱합으로 계산, 평균을 내어 손실 함수를 구하는 방식.

오차가 더 큼에도 불구하고 더 좋은 평가를 받아 학습이 느려지거나 학습이 안 되는 상황이 발생하기도 한다.

## MAE

> Mean Absolute Error

오차가 작아도 선형적으로 오차를 평가하고 싶을 때 절대값으로 계산한다

## 후버 손실

MSE와 MAE의 장점을 결합하기 위해 작은 오차는 MSE를, 큰 오차는 MAE를 사용하는 함수.

### CODE REVIEW

**Comment by lego37yoon** [comment]
아직 준비 중입니다

**Comment by kimhamyong** [comment]
task1:

```python
# 기존 모델에서 hidden1부터 hidden3까지를 가져와 연결
hidden1 = base_model.get_layer("hidden1")
hidden2 = base_model.get_layer("hidden2")
hidden3 = base_model.get_layer("hidden3")
```
이 부분은 기존 모델에서 hidden1, hidden2, hidden3 레이어를 단순히 가져오게 되어 새로운 입력 레이어와 연결되지 않으므로 데이터 흐름이 설정되지 않습니다.

```python
hidden1 = base_model.get_layer('hidden1')(input_layer)
hidden2 = base_model.get_layer('hidden2')(hidden1)
hidden3 = base_model.get_layer('hidden3')(hidden2)
```
이러한 방식으로  Functional API를 사용해 새로운 입력 레이어(input_layer)와 기존 레이어를 명시적으로 연결하여 데이터 흐름이 input_layer → hidden1 → hidden2 → hidden3로 설정되도록 해야합니다.

**Comment by kimhamyong** [comment]
task1:

```python
[(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)]
```
기존 코드에서는 데이터 전처리를 제대로 수행하지 않아 몇 가지 문제가 있습니다.

- 데이터가 [0, 255] 범위의 정수형(uint8)으로 유지되므로 딥러닝 모델의 부동소수점 연산(float32)에 적합하지 않습니다. 이로 인해 추가적인 연산 비용이 발생하거나 데이터 타입 불일치로 모델 학습의 안정성이 저하될 수 있습니다.
- 픽셀 값을 [0, 1]로 정규화하지 않아 입력 데이터의 범위가 크므로 활성화 함수에서 포화가 발생할 가능성이 있습니다.
- reshape(60000, 784)와 같은 하드코딩 방식은 데이터셋의 크기가 변할 경우 코드 수정이 필요하게 됩니다.


```python
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
x_train = x_train.reshape(-1, 784).astype('float32') / 255.0  # 데이터 타입 변경 및 정규화
x_test = x_test.reshape(-1, 784).astype('float32') / 255.0    # 데이터 타입 변경 및 정규화
```
따라서 이러한 방법을 사용하여 개선할 수 있습니다.

- 데이터를 float32 타입으로 변환해 모델 연산에 적합하게 만듭니다.
- [0, 1]로 정규화하여 학습 속도와 안정성을 높이며, 활성화 함수의 포화를 방지합니다.
-  reshape(-1, 784)를 사용해 샘플 수를 자동으로 계산하므로 다른 데이터셋에서도 유연하게 사용할 수 있습니다.

**Comment by kimhamyong** [comment]
task1:

테스트 정확도가 64.40%로 낮고 초기 학습 손실이 52.2743으로 비정상적으로 크게 나타나는 것은 보면 학습 과정에서 그래디언트 폭주 문제가 발생했을 가능성이 있습니다. 그래디언트 폭주는 특정 레이어에서 그래디언트 값이 지나치게 커져 가중치 업데이트가 제대로 이루어지지 않는 현상으로 아래와 같은 개선을 적용할 수 있습니다.

1. 그래디언트의 크기를 제한합니다. 이를 위해 옵티마이저에 clipnorm=1.0을 설정하여 그래디언트 벡터의 크기가 1을 넘지 않도록 제한합니다.
```python
transfer_model.compile(
    optimizer=Adam(learning_rate=0.001, clipnorm=1.0),
    loss=SparseCategoricalCrossentropy(),
    metrics=['accuracy']
)
```

2. 학습률이 너무 크면 학습이 불안정해질 수 있으므로 학습이 진행됨에 따라 학습률을 점진적으로 줄입니다. Keras의 ExponentialDecay를 사용해 초기 학습률을 설정하고 단계적으로 감소시킵니다.
```python
lr_schedule = ExponentialDecay(
    initial_learning_rate=0.001,
    decay_steps=1000,
    decay_rate=0.9
)

transfer_model.compile(
    optimizer=Adam(learning_rate=lr_schedule, clipnorm=1.0),
    loss=SparseCategoricalCrossentropy(),
    metrics=['accuracy']
)
```

3. 추가된 레이어(hidden4, logits)의 가중치 초기화가 적절하지 않으면 학습 초기 손실이 과도하게 커질 수 있습니다. ReLU 활성화 함수에는 he_normal, 소프트맥스 출력 레이어에는 glorot_uniform 초기화를 사용합니다.
```python
hidden4 = Dense(16, activation='relu', kernel_initializer='he_normal', name='hidden4')
logits = Dense(10, activation='softmax', kernel_initializer='glorot_uniform', name='logits')
```

**Review by giyomdabin — APPROVED** [review]
task2: 주어진 과제를 잘 수행하셨습니다.

손실 값의 기복 문제와 특정 샘플의 큰 오차를 해결하기 위해서는 다음 방법을 고려할 수 있습니다:

- 은닉층에 Batch Normalization을 추가하면 학습 과정이 더욱 안정적으로 이루어질 수 있습니다. 이는 각 층의 입력 분포를 정규화하여 손실 값의 기복을 줄이고, 학습률 설정에 덜 민감하게 만듭니다.
- 과적합 방지와 더불어 학습 과정의 안정성을 높이기 위해 Dropout을 추가할 수 있습니다.
- 학습률이 고정되어 있으면 초기 학습 단계에서는 너무 크게, 후반부에는 너무 작게 작용할 수 있습니다. 이를 방지하기 위해 학습률 감소(Learning Rate Decay)를 도입할 수 있습니다.

또한, MSE 외에 MAE(Mean Absolute Error)를 추가로 계산하면 모델 성능을 다각도로 평가할 수 있습니다.

- MSE는 **큰 오차(Outlier)**에 민감한 지표입니다. 따라서 MSE만을 기준으로 모델 성능을 평가하면 극단적인 오차에 영향을 많이 받아, 모델의 평균적인 성능이 과소평가될 가능성이 있습니다.
- MAE는 큰 오차에 덜 민감하며, 모든 데이터에 걸쳐 평균적인 오차 크기를 직접적으로 보여줍니다.

따라서 두 지표를 함께 분석하면 모델 성능을 더 다각적으로 평가할 수 있으며, 특정 데이터의 특성에 맞게 개선 방향을 도출할 수 있습니다.

수고하셨습니다.

**Review by kimhamyong — APPROVED** [review]
task1: 잘 수행하셨습니다.

마지막 과제까지 수고 많으셨습니다!
